typed-function.js

```diff
@@ -1259,96 +1259,137 @@
+
+     * The name of the resulting typed-function will be given by the
+     * string-valued name argument if present, or if not, by the name
+     * of any of the arguments that have one, as long as any that do are
+     * consistent with each other. If no name is specified, the name will be
+     * an empty string.
+     *
+     * @param {string} name [optional]
+     * @param {(function|object)[]} signature providers
+     * @returns {typed-function}
+     */
+    typed = function(maybeName) {
+      const named = typeof maybeName === 'string'
+      let name = named ? maybeName : ''
+      let start = named ? 1 : 0
+      let allSignatures = {}
+      for (let i = start; i < arguments.length; ++i) {
+        const item = arguments[i]
+        let theseSignatures
+        let thisName
+        if (typeof item === 'function') {
+          thisName = item.name
+          /* Case 1: a single function with a string signature property */
+          if (typeof item.signature === 'string') {
+            theseSignatures = {}
+            theseSignatures[item.signature] = item
+          }
+          /* Case 2: an existing typed function */
+          else if (typeof item.signatures === 'object') {
+            theseSignatures = item.signatures
+          } else {
+            /* Oops, not a function we can deal with */
+            const err = new TypeError(
+              "Cannot merge a function without a 'signature' property " +
+              '(index ' + i + ', named: ' + thisName + ')')
+            err.data = { index: i, name: thisName }
+            throw err
           }
         }
-        var name = getName(fns);
-        return createTypedFunction(name, signaturesMap);
-      },
-      '...Function': function (fns) {
-        return createTypedFunction(getName(fns), extractSignatures(fns));
-      },
-      'string, ...Function': function (name, fns) {
-        return createTypedFunction(name, extractSignatures(fns));
+        /* Case 3: Plain object, with keys = signatures, values = functions */
+        else if (typeof item === 'object' &&
+                 (!(item.constructor) || item.constructor.name === 'Object')) {
+          theseSignatures = item
+          if (!named) {
+            thisName = getObjectName(item)
+          }
+        } else {
+          /* Oops, not a type we can deal with */
+          const err = new TypeError(
+            'Can only create typed-functions from functions, or objects ' +
+            'with signatures as keys and functions as values (not "' +
+            item.toString() + '" at index ' + i + ')')
+          err.data = { index: i, item: item }
+          throw err
+        }
+        /* Item is ok and we have the signatures and maybe a name */
+        if (!named) {
+          name = checkName(name, thisName)
+        }
+        mergeSignatures(allSignatures, theseSignatures)
       }
-    });
+      return createTypedFunction(name || '', allSignatures)
```

Comment:
I'm not sure about setting the default name to an empty string `''`, is that indeed needed? Or does it make sense to leave it undefined, and handle the undefined case when it occurs?

Reply:
Well, I originally had just `name` there, but the unit tests assert that all unnamed typed functions have precisely the name `''` so for them to pass I had to change it to this expression `name | ''`. Let me know if you'd rather I change the tests to match the behavior without the `|| ''`, or leave things as they are now.

Reply:
Ah ok, makes sense to keep it like this then 👍 .